source/math/Quaternion.ooc

```diff
@@ -260,6 +263,99 @@ Quaternion: cover {
+		// TODO: Fix singularity problem when the angle is close to PI, using sequential rotations
+		(referenceVectors, observationVectors) := This _createVectorMeasurementsForQuest(quaternions)
+		normalizedWeights := weights / weights sum
+
+		attitudeProfile := FloatMatrix new(3, 3) take()
+		for (currentVector in 0 .. referenceVectors width)
+			attitudeProfile += normalizedWeights[currentVector / 3] / 3.0f * observationVectors getColumn(currentVector) * referenceVectors getColumn(currentVector) transpose()
+		matrixQuantityS := (attitudeProfile + attitudeProfile transpose()) take()
+
+		vectorQuantityZ := FloatMatrix new(1, 3) take()
+		temporaryZ := FloatPoint3D new()
+		for (currentVector in 0 .. referenceVectors width) {
+			currentObservationVector := FloatPoint3D new(observationVectors[currentVector, 0], observationVectors[currentVector, 1], observationVectors[currentVector, 2])
+			currentReferenceVector := FloatPoint3D new(referenceVectors[currentVector, 0], referenceVectors[currentVector, 1], referenceVectors[currentVector, 2])
+			temporaryZ += normalizedWeights[currentVector / 3] / 3.0f * currentObservationVector vectorProduct(currentReferenceVector)
+		}
+		vectorQuantityZ setVertical(0, 0, temporaryZ)
+
+		maximumEigenvalue := This _approximateMaximumEigenvalueForQuest(matrixQuantityS, vectorQuantityZ, 1.0f, 5)
+		linearCoefficients := ((maximumEigenvalue + attitudeProfile trace()) * FloatMatrix identity(3) - matrixQuantityS)
+		gibbsVector := linearCoefficients solve(vectorQuantityZ)
+
+		gibbsVectorSquaredNorm := 0.0f
+		for (index in 0 .. gibbsVector height)
+			gibbsVectorSquaredNorm += gibbsVector[0, index] squared()
+		vector := FloatMatrix new(1, 4)
+		vector[0, 3] = 1.0f
+		for (index in 0 .. 3)
+			vector[index, 0] = gibbsVector[index, 0]
+		quaternionValues := (1.0f / sqrt(1.0f + gibbsVectorSquaredNorm)) * vector
+		result := This new(quaternionValues[3, 0], -quaternionValues[0, 0], -quaternionValues[1, 0], -quaternionValues[2, 0])
+
+		normalizedWeights free()
+		referenceVectors free()
+		observationVectors free()
+		attitudeProfile free()
+		matrixQuantityS free()
+		vectorQuantityZ free()
+		gibbsVector free()
+		quaternionValues free()
+		result
+	}
+	_approximateMaximumEigenvalueForQuest: static func (matrixQuantityS, vectorQuantityZ: FloatMatrix, initialGuess: Float, maximumIterationCount := 20) -> Float {
+		sigma := 0.5f * matrixQuantityS trace()
+		constantA := sigma squared() - matrixQuantityS adjugate() trace()
+		temporary := vectorQuantityZ transpose() * vectorQuantityZ
+		constantB := sigma squared() + temporary[0, 0]
+		temporary free()
+		temporary = vectorQuantityZ transpose() * matrixQuantityS * vectorQuantityZ
+		constantC := matrixQuantityS determinant() + temporary[0, 0]
+		temporary free()
+		temporary = vectorQuantityZ transpose() * matrixQuantityS * matrixQuantityS * vectorQuantityZ
+		constantD := temporary[0, 0]
+		temporary free()
+
+		for (_ in 0 .. maximumIterationCount) {
```

Comment:
Haha, yes, does it look weird? I thought it made sense since I am only interested in limiting the number of iterations, not to actually access the iteration variable.